flake8_trio/visitors/helpers.py

```diff
@@ -174,3 +189,63 @@ def get_matching_call(
     ):
         return node, node.func.attr, node.func.value.id
     return None
+
+
+# ___ CST helpers ___
+def oneof_names(*names: str):
+    return m.OneOf(*map(m.Name, names))
+
+
+CSTNode_T = TypeVar("CSTNode_T", bound=cst.CSTNode)
+
+
+def list_contains(
+    seq: Sequence[CSTNode_T], matcher: m.BaseMatcherNode
+) -> Iterator[CSTNode_T]:
+    yield from (item for item in seq if m.matches(item, matcher))
+
+
+class AttributeCall(NamedTuple):
+    node: cst.Call
+    base: str
+    function: str
+
+
+def with_has_call(
+    node: cst.With, *names: str, base: Iterable[str] = ("trio", "anyio")
+) -> AttributeCall | None:
+    for item in node.items:
+        if res := m.extract(
+            item.item,
+            m.Call(
+                func=m.Attribute(
+                    value=m.SaveMatchedNode(m.Name(), name="library"),
+                    attr=m.SaveMatchedNode(
+                        oneof_names(*names),
+                        name="function",
+                    ),
+                )
+            ),
+        ):
+            assert isinstance(item.item, cst.Call)
+            assert isinstance(res["library"], cst.Name)
+            assert isinstance(res["function"], cst.Name)
+            if res["library"].value not in base:
+                return None
+            return AttributeCall(item.item, res["library"].value, res["function"].value)
+    return None
```

Comment:
I think here we actually want to continue looping if it's an unknown base?  More generally I have a (weak) preference to surface all the errors we can find in a single shot, and so I'd be inclined to return a `list[AttributeCall]`.

Reply:
Oops yes, it should very much continue looping.

Reply:
and yep, changed it to return a list.